### PR TITLE
[mlir][vector][xegpu] Vector to XeGPU conversion pass

### PR DESCRIPTION
**Pull request by adam-smnk**
Add pass for Vector to XeGPU dialect conversion and initial conversion patterns for vector.transfer_read|write operations.

### CODE REVIEW

File: mlir/lib/Conversion/VectorToXeGPU/VectorToXeGPU.cpp [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+  using OpRewritePattern<vector::TransferWriteOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(vector::TransferWriteOp writeOp,
+                                PatternRewriter &rewriter) const override {
+    Location loc = writeOp.getLoc();
+
+    if (failed(transferPreconditions(rewriter, writeOp)))
+      return failure();
+
+    if (writeOp.hasOutOfBoundsDim())
```
Comment (adam-smnk) [inline mlir/lib/Conversion/VectorToXeGPU/VectorToXeGPU.cpp:207]: @chencha3 Actually I'm not sure if out of bound `store_nd` is supported

**Comment by adam-smnk** [comment]
@chencha3 ping

**Comment by chencha3** [comment]
> @chencha3 ping

Sorry, missed it. I am looking at it.

File: mlir/lib/Conversion/VectorToXeGPU/VectorToXeGPU.cpp (hunk 1 above)
Comment (chencha3) [reply mlir/lib/Conversion/VectorToXeGPU/VectorToXeGPU.cpp:207]: It depends on the hardware, I think. Per PVC spec, it is. "For loads, hardware will return 0 for the out-of-bound bytes. For stores, hardware will not update the out-of-bound bytes in memory"

**Review by chencha3 — APPROVED** [review]
LGTM

**Comment by kazutakahirata** [comment]
I'm getting a couple of errors of the form:

```
mlir/lib/Conversion/VectorToXeGPU/VectorToXeGPU.cpp:172:16: error: no member named 'MemoryScope' in namespace 'mlir::xegpu'
        xegpu::MemoryScope::Global,
        ~~~~~~~^
```

Would mind taking a look?  Thanks!

**Comment by adam-smnk** [comment]
@kazutakahirata This wasn't updated after changes in #109144, revert pending #109666 
@chencha3 After rebase and tweaks, both should be fine

**Comment by chencha3** [comment]
> @kazutakahirata This wasn't updated after changes in #109144, revert pending #109666 @chencha3 After rebase and tweaks, both should be fine

Yeah, rebased it. It should be fixed now.

**Comment by kazutakahirata** [comment]
> Yeah, rebased it. It should be fixed now.

Thanks a lot!

**Pull request merged**